Added specification for the WAN pairing endpoint registry (#76)
flexiblepower/s2-documentation: website/s2-connect/discovery-pairing-authentication.md

| Part of specification | Description | Specification format | Location |
| --- | --- | --- | --- |
| S2 Connect pairing API | HTTP based interaction to pair two nodes | OpenAPI file | [Github](https://github.com/flexiblepower/s2-connect/blob/main/s2-connect-pairing.yml) |
| S2 Connect connection API | HTTP based interaction set up a communication channel for S2 messages between two nodes | OpenAPI file | [Github](https://github.com/flexiblepower/s2-connect/blob/main/s2-connect-connection-init.yml) |
| S2 Connect WAN pairing endpoint registry API | HTTP based interface to query the registry  | OpenAPI file | [Github](https://github.com/flexiblepower/s2-connect/blob/main/s2-connect-wan-endpoint-registry.yml) |
| S2 JSON message structure | The types of S2 messages that can be exchanges between nodes | JSON schema files | [Github](https://github.com/flexiblepower/s2-ws-json/tree/main/s2-json-schema) |

## Versioning of OpenAPI files

> The S2 Connect pairing API, the S2 Connect connection initiation API and the S2 Connect WAN pairing endpoint registry are formally defined in OpenAPI files. To accommodate future changes to these APIs, the OpenAPI files are versioned. Versioning is done using a `major.minor` scheme. All S2 Connect OpenAPI files share the same version number.

Reviewer: To have textual symmetry, add the word API:
The S2 Connect pairing API, the S2 Connect connection initiation API and the S2 Connect WAN pairing endpoint registry **API**

Reviewer: Since this is the S2 Connect specification, I have a slight preference for using: The pairing API, the connection initiation API and the WAN pairing endpoint registry API are formally...

## Discovery

> In order to ease the pairing process, which is specified below, the discovery process provides a way for nodes to find each other without requiring a user to know the pairing endpoint of the other node. In other words, the discovery process is a way to provide an node with the URL of another node which is needed to start the pairing process. Alternatively, it **must** always possible to initiate the pairing by manually providing the URL by the end user.

Reviewer: ```suggestion
In order to ease the pairing process, which is specified below, the discovery process provides a way for nodes to find each other without requiring a user to know the pairing endpoint URL of the other node. In other words, the discovery process is a way to provide a node with the URL of another node which is needed to start the pairing process. Alternatively, it **must** always be possible to initiate the pairing by manually providing the URL by the end user.
```

> There are two mechanisms for discovery: For discovering WAN endpoints there is a central online registry. For discovering endpoints within the same LAN, DNS-SD is used. DNS-SD can also be used to discovery WAN endpoints of devices that have a presence in the LAN (typically an energy smart appliance which hosts the RM in the cloud).

Reviewer: ```suggestion
There are two mechanisms for discovery: For discovering WAN endpoints there is a central online registry. For discovering endpoints within the same LAN, DNS-SD is used. DNS-SD can also be used to discovery WAN endpoints of devices that have a presence in the LAN (typically an energy smart appliance of which the RM is deployed in the cloud).
```

### WAN pairing endpoint registry

> Note: At this point the registry is specified, but not yet publicly available

> The task of the registry is to facilitate a more user friendly way to determine the URL of the WAN pairing endpoint. Owners of an S2 Connect WAN pairing endpoint can register their endpoint at the registry. The user interface of a CEM or RM could show a list of relevant endpoints to the user (e.g. in a list or drop down menu) with details that would be easily recognizable to the end user (e.g. name and icon). By querying the registry the user interface can always show an up-to-date list of endpoints. The registry contains filtering functionality to filter endpoints that are relevant in the context.

Reviewer: ```suggestion
The purpose of the registry is to facilitate a more user friendly way to determine the URL of the WAN pairing endpoint. Providers of an S2 Connect WAN pairing endpoint can register their endpoint at the registry. The user interface of a CEM or RM could show a list of relevant endpoints to the user (e.g., in a list or drop down menu) with details that would be easily recognizable to the end user (e.g., name and icon). By querying the registry, the user interface can always show an up-to-date list of endpoints. The registry contains filtering functionality to filter endpoints that are relevant in the context.
```

> The registry uses the same versioning mechanism as the other S2 Connect OpenAPI files. See [Selecting the version of REST APIs](#selecting-the-version-of-rest-apis) to see how the client can select the version of the API to use.

Reviewer: ```suggestion
The registry uses the same version negotiation mechanism as the other S2 Connect OpenAPI files. Refer to [Selecting the version of REST APIs](#selecting-the-version-of-rest-apis) for information on how clients can select the API version to use.
```

| `name` | User facing name of the endpoint |
| `description` | User facing description of the endpoint|
| `icon32` | 32 by 32 pixel- icon of the endpoint |
| `icon128` | 128 by 128 pixels icon of the endpoint |
| `icon512` | 512 by 512 pixels icon of the endpoint |
| `pairingApiUrl` | The URL of the pairing API of the endpoint |
| `regions` | Array of regions in which this endpoint operates, as defined by the ISO 3166-1 alpha-2 country code |
| `status` | Status of the endpoint, can either be `testing` or `public` |
| `cem` | Boolean indicating if the endpoint represents CEM nodes |
| `rm` | Boolean indicating if the endpoint represents RM nodes |

In order to filter out the relevant endpoint records the API supports the following filters:

* Region in which the endpoint claims to be active
* Status, which can be `testing` or `public` (when no value provided, the default value will be `public`)
* Whether the endpoint contains CEM nodes
* Whether the endpoint contains RM nodes

In addition, the number of responses can be limited. An offset can also be provided in order to split the results over multiple requests.

### DNS-SD based discovery

> DNS-SD is used for automatically discover nodes from an node that is deployed in the LAN. This method can be used in three ways.

Reviewer: ```suggestion
DNS-SD is used to automatically discover nodes from a node that is deployed in the LAN. This method can be used in three ways.
```

* To discover another node that is deployed in the LAN, which is the responder node
* To advertise a [long polling URL](#long-polling) so other initiator nodes in the LAN could connect to this node

> * To discover an node of which the RM is deployed in the WAN, but that also has a presence in the LAN.

Reviewer: ```suggestion
* To discover a node of which the RM is deployed in the WAN, but that also has a presence in the LAN.
```
